src/qdmi/Driver.cpp

```diff
@@ -0,0 +1,479 @@
+    LOAD_STATIC_SYMBOL(prefix, device_session_alloc)                           \
+    LOAD_STATIC_SYMBOL(prefix, device_session_init)                            \
+    LOAD_STATIC_SYMBOL(prefix, device_session_free)                            \
+    LOAD_STATIC_SYMBOL(prefix, device_session_set_parameter)                   \
+    /* device job interface */                                                 \
+    LOAD_STATIC_SYMBOL(prefix, device_session_create_device_job)               \
+    LOAD_STATIC_SYMBOL(prefix, device_job_free)                                \
+    LOAD_STATIC_SYMBOL(prefix, device_job_set_parameter)                       \
+    LOAD_STATIC_SYMBOL(prefix, device_job_query_property)                      \
+    LOAD_STATIC_SYMBOL(prefix, device_job_submit)                              \
+    LOAD_STATIC_SYMBOL(prefix, device_job_cancel)                              \
+    LOAD_STATIC_SYMBOL(prefix, device_job_check)                               \
+    LOAD_STATIC_SYMBOL(prefix, device_job_wait)                                \
+    LOAD_STATIC_SYMBOL(prefix, device_job_get_results)                         \
+    /* device query interface */                                               \
+    LOAD_STATIC_SYMBOL(prefix, device_session_query_device_property)           \
+    LOAD_STATIC_SYMBOL(prefix, device_session_query_site_property)             \
+    LOAD_STATIC_SYMBOL(prefix, device_session_query_operation_property)        \
+    /* NOLINTEND(cppcoreguidelines-pro-type-reinterpret-cast) */               \
+    /* initialize the device */                                                \
+    device_initialize();                                                       \
+  }                                                                            \
+                                                                               \
+  prefix##DeviceLibrary::~prefix##DeviceLibrary() {                            \
+    /* Check if QDMI_device_finalize is not NULL before calling it. */         \
+    if (device_finalize != nullptr) {                                          \
+      device_finalize();                                                       \
+    }                                                                          \
+  }
+DEFINE_STATIC_LIBRARY(MQT_NA)
+
+#ifndef _WIN32
+DynamicDeviceLibrary::DynamicDeviceLibrary(const std::string& libName,
+                                           const std::string& prefix)
+    : libHandle_(dlopen(libName.c_str(), RTLD_NOW | RTLD_LOCAL)) {
+  if (libHandle_ == nullptr) {
+    throw std::runtime_error("Couldn't open the device library: " + libName);
+  }
+  if (openLibHandles().contains(libHandle_)) {
+    // dlopen uses reference counting, so we need to decrement the reference
+    // count that was increased by dlopen.
+    dlclose(libHandle_);
+    throw std::runtime_error("Device library already loaded: " + libName);
+  }
+  openLibHandles().emplace(libHandle_);
```

Comment:
I think this can be slightly more idiomatic by using `try_emplace`, which combines the check with the emplacement in case it is not present and gives you a success indicator as a return value.